private/tools/java/com/github/bazelbuild/rules_jvm_external/maven/MavenPublisher.java

```diff
@@ -82,101 +86,136 @@
+          publishMavenMetadata,
+          args.length > 4 ? args[4] : null,
+          repo,
+          credentials,
+          signingMetadata,
+          executorService);
+    } finally {
+      executorService.shutdown();
     }
+  }
 
-    Coordinates coords = new Coordinates(parts.get(0), parts.get(1), parts.get(2));
+  protected static void run(
+      String coordinates,
+      String pomPath,
+      String mainArtifactPath,
+      boolean publishMavenMetadata,
+      @Nullable String extraArtifacts,
+      String repo,
+      @Nullable Credentials credentials,
+      SigningMetadata signingMetadata,
+      Executor executor)
+      throws Exception {
 
-    // Calculate md5 and sha1 for each of the inputs
-    Path pom = Paths.get(args[1]);
-    Path mainArtifact = getPathIfSet(args[2]);
+    if (!isSchemeSupported(repo)) {
+      throw new IllegalArgumentException(
+          "Repository must be accessed via the supported schemes: "
+              + Arrays.toString(SUPPORTED_SCHEMES));
+    }
 
-    try {
-      List<CompletableFuture<Void>> futures = new ArrayList<>();
-      futures.add(upload(repo, credentials, coords, ".pom", pom, signingMetadata));
+    if (!isUploadSchemeSupported(repo)) {
+      throw new IllegalArgumentException(
+          "Repository must be uploaded to via the supported schemes: "
+              + Arrays.toString(SUPPORTED_UPLOAD_SCHEMES));
+    }
 
-      if (mainArtifact != null) {
-        String ext =
-            com.google.common.io.Files.getFileExtension(mainArtifact.getFileName().toString());
-        futures.add(upload(repo, credentials, coords, "." + ext, mainArtifact, signingMetadata));
-      }
+    final Coordinates coords = Coordinates.fromString(coordinates);
 
-      if (args.length > 4 && !args[4].isEmpty()) {
-        List<String> extraArtifactTuples = Splitter.onPattern(",").splitToList(args[4]);
-        for (String artifactTuple : extraArtifactTuples) {
-          String[] splits = artifactTuple.split("=");
-          String classifier = splits[0];
-          Path artifact = Paths.get(splits[1]);
-          String ext = com.google.common.io.Files.getFileExtension(splits[1]);
-          futures.add(
-              upload(
-                  repo,
-                  credentials,
-                  coords,
-                  String.format("-%s.%s", classifier, ext),
-                  artifact,
-                  signingMetadata));
-        }
+    // Calculate md5 and sha1 for each of the inputs
+    Path pom = Paths.get(pomPath);
+
+    List<CompletableFuture<Void>> futures = new ArrayList<>();
+    futures.add(upload(repo, credentials, coords, ".pom", pom, signingMetadata, executor));
+
+    futures.add(
+        upload(
+            repo,
+            credentials,
+            coords,
+            "." + getFileExtension(mainArtifactPath),
+            Paths.get(mainArtifactPath),
+            signingMetadata,
+            executor));
+
+    if (!Strings.isNullOrEmpty(extraArtifacts)) {
```

Comment:
empty string being passed for extraArtifacts was causing an IndexOutOfBounds exception